SURF-768 Documentation for `cardinality` function (#1574)
neo4j/docs-cypher: modules/ROOT/pages/functions/scalar.adoc

======

[role=label--new-Neo4j-2026.06 label--cypher-25-only]
[[functions-cardinality]]
== cardinality()

.Details
|===
| *Syntax* 3+| `cardinality(input)`
| *Description* 3+| Returns the number of keys in a `MAP`, the number of elements in a `LIST<ANY>`, or the number of nodes and relationships in a `PATH`.
.2+| *Arguments* | *Name* | *Type* | *Description*
| `input` | `MAP \| LIST<ANY> \| PATH` | A list, path or map whose cardinality is to be calculated.
| *Returns* 3+| `INTEGER`
|===

- For a `MAP`, it returns the same result as `size(keys(map))`.
- For a `LIST<ANY>`, `cardinality()` returns the same result as xref:functions/scalar.adoc#functions-size[`size()`].
- For a `PATH`, it returns the total number of nodes and relationships in the path.

> - For a non-empty path, `cardinality(p) = 2 * length(p) + 1`.

Reviewer: This could go under Considerations I think.

.Considerations
|===

| `cardinality(null)` returns `null`.
| For a `LIST<ANY>`, only top-level elements are counted; nested lists count as a single element.

|===

[NOTE]

> To calculate the length of a `PATH` in terms of traversed relationships only, see xref:functions/scalar.adoc#functions-length[`length()`] or xref:functions/scalar.adoc#functions-path-length[`path_length()`].

Reviewer: ```suggestion
To calculate the length of a `PATH` as a relationship count only, see xref:functions/scalar.adoc#functions-length[`length()`] or xref:functions/scalar.adoc#functions-path-length[`path_length()`].
```

======

.Query
// tag::functions_scalar_cardinality_list[]
[source, cypher, indent=0, role=test-skip]
----
RETURN cardinality([1, 2, 3]) AS result
----
// end::functions_scalar_cardinality_list[]

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===

| result
| 3
1+d|Rows: 1

|===

The number of elements in the list is returned.

======

.+cardinality() applied to maps+
======

.Query
// tag::functions_scalar_cardinality_map[]
[source, cypher, indent=0, role=test-skip]
----
RETURN cardinality({a: 1, b: 2}) AS result,
       cardinality({a: 1, b: 2}) = size(keys({a: 1, b: 2})) AS sameAsSizeKeys
----
// end::functions_scalar_cardinality_map[]

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===

| result | sameAsSizeKeys
| 2 | true
2+d|Rows: 1

|===

The number of keys in the map is returned.

======

.+cardinality() applied to paths+
======

.Query
// tag::functions_scalar_cardinality_path[]
[source, cypher, indent=0, role=test-skip]
----
MATCH p = (a:Developer)-[:KNOWS]->(b:Administrator)
WHERE a.name = 'Alice' AND b.name = 'Bob'
RETURN cardinality(p) AS cardinality, length(p) AS length
----
// end::functions_scalar_cardinality_path[]

> The cardinality and length of the path `p` are returned.

Reviewer: ```suggestion
The cardinality of the returned path is `3` (2 nodes and 1 relationship), while its length is `1` (for the single relationship traversed).
```

======

[role=label--new-Neo4j-2026.06 label--cypher-25-only]
[[functions-cardinality]]
== cardinality()

.Details
|===
| *Syntax* 3+| `cardinality(input)`
| *Description* 3+| Returns the number of keys in a `MAP`, the number of elements in a `LIST<ANY>`, or the number of nodes and relationships in a `PATH`.
.2+| *Arguments* | *Name* | *Type* | *Description*

> | `input` | `LIST<ANY> \| MAP \| PATH` | A list, path or map whose cardinality is to be calculated.

Reviewer 2: ```suggestion
| `input` | `MAP \| LIST<ANY> \| PATH` | A list, path or map whose cardinality is to be calculated.
```

Please keep the order to that which is normalised by Cypher (e.g what is printed in the signature of the function)

| *Returns* 3+| `INTEGER`
|===

> This function was introduced as part of Cypher's xref:appendix/gql-conformance/index.adoc[] (optional GQL feature GF12, "CARDINALITY function").

Reviewer 2: Not sure if this is needed, it offers new functionality unlike the alias ones you did, maybe Jens has an opinon on that though

Reviewer: Agreed - this is not necessary here.

- For a `LIST<ANY>`, `cardinality()` returns the same result as xref:functions/scalar.adoc#functions-size[`size()`].

> - For a `MAP`, it returns the same result as `size(keys(map))`.

Reviewer 2: Can you switch the order to match the description?

- For a `PATH`, it returns the total number of nodes and relationships in the path.
- For a non-empty path, `cardinality(p) = 2 * length(p) + 1`.

> Binding table reference values are not supported in Cypher, so this is an intentional subset of the full GQL cardinality expression.

Reviewer 2: This is not relevant here, it is most likely confusing to users who don't know GQL well

.Considerations
|===

| `cardinality(null)` returns `null`.
| For a `LIST<ANY>`, only top-level elements are counted; nested lists count as a single element.

> | `cardinality()` on values other than a `LIST<ANY>`, `MAP`, `PATH`, or `null` will fail the query.

Reviewer 2: This is true of all functions, I don't think we normally say that
